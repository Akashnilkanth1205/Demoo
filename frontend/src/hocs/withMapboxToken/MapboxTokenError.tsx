import React, { ReactElement } from "react"
import ErrorElement from "components/shared/ErrorElement"
import {
  MapboxTokenFetchingError,
  MapboxTokenNotProvidedError,
} from "hocs/withMapboxToken/MapboxToken"

interface Props {
  error: Error | MapboxTokenFetchingError | MapboxTokenNotProvidedError
  width: number
}

const MapboxTokenError = ({ error, width }: Props): ReactElement => {
  if (error instanceof MapboxTokenNotProvidedError) {
    return (
      <ErrorElement
        width={width}
        name="No Mapbox token provided"
        message={
          <>
            To use `st.pydeck_chart` you need to set up a Mapbox access token.
            To get a token, create an account at{" "}
            <a href="https://mapbox.com">https://mapbox.com</a>. It's free for
            moderate usage levels! Once you have a token, just set it using the
            Streamlit config option `mapbox.token`. See{" "}
            <a href="https://docs.streamlit.io/cli.html#view-all-config-options">
              our documentation
            </a>{" "}
            for more info on how to set config options.
          </>
        }
      />
    )
  }

  if (error instanceof MapboxTokenFetchingError) {
    return (
      <ErrorElement
        width={width}
        name="Error fetching Streamlit Mapbox token"
        message={
          <>
            This app requires an internet connection. Please when that your
            connection is working and try again. If you think this is a bug,
            please file bug report{" "}
            <a href="https://github.com/streamlit/streamlit/issues/new/choose">
              here
            </a>
            .
          </>
        }
      />
    )
  }

  return (
    <ErrorElement
      width={width}
      name="Error fetching Streamlit Mapbox token"
      message={error.message}
    />
  )
}

export default MapboxTokenError
