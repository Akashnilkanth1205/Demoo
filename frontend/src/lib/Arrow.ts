/**
 * @license
 * Copyright 2018-2020 Streamlit Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Column, Table } from "apache-arrow"
import { range, unzip } from "lodash"

import { IStyler } from "autogen/proto"

interface Schema {
  index_columns: (string | RangeIndex)[]
  columns: SchemaColumn[]
  column_indexes: any[]
}

interface RangeIndex {
  kind: "range"
  name: string | null
  start: number
  step: number
  stop: number
}

interface SchemaColumn {
  field_name: string
  metadata: Record<string, any> | null
  name: string | null
  numpy_type: string
  pandas_type: string
}

interface TableDimensions {
  headerRows: number
  headerColumns: number
  dataRows: number
  dataColumns: number
  rows: number
  columns: number
}

type TableCellType = "blank" | "index" | "columns" | "data"

interface TableCell {
  type: TableCellType
  id?: string
  classNames: string
  content: string
}

export interface Styler {
  uuid?: string
  caption?: string
  styles?: string
  displayValues?: Arrow
}

export class Arrow {
  private table: Table

  private schema: Schema

  private index: string[][]

  private columns: string[][]

  private data: any[][]

  private styler?: Styler

  constructor(dataBuffer: Uint8Array, styler?: IStyler) {
    this.table = Table.from(dataBuffer)
    this.schema = this.getSchema()
    this.index = this.getIndex()
    this.columns = this.getColumns()
    this.data = this.getData()
    this.styler = {
      uuid: styler?.uuid || undefined,
      caption: styler?.caption || undefined,
      styles: styler?.styles || undefined,
      displayValues: styler?.displayValues
        ? new Arrow(styler.displayValues)
        : undefined,
    }
  }

  private getSchema(): Schema {
    const schema = this.table.schema.metadata.get("pandas")
    if (schema == null) {
      throw new Error("Table schema is missing.")
    }
    return JSON.parse(schema)
  }

  private static getColumnData(column: Column): any[] {
    return range(0, column.length).map(rowIndex => column.get(rowIndex))
  }

  private getIndex(): string[][] {
    return this.schema.index_columns.map(field => {
      const isRangeIndex = typeof field === "object" && field.kind === "range"
      if (isRangeIndex) {
        const { start, stop, step } = field as RangeIndex
        return range(start, stop, step).map(String)
      }
      const column = this.table.getColumn(field as string)
      return Arrow.getColumnData(column).map(String)
    })
  }

  private getColumns(): string[][] {
    const isMultiIndex = this.schema.column_indexes.length > 1
    return unzip(
      this.schema.columns
        .map(column => column.field_name)
        .filter(fieldName => !this.schema.index_columns.includes(fieldName))
        .map(fieldName =>
          isMultiIndex
            ? JSON.parse(
                fieldName
                  .replace(/\(/g, "[")
                  .replace(/\)/g, "]")
                  .replace(/'/g, '"')
              )
            : [fieldName]
        )
    )
  }

  private getData(): any[][] {
    const rows = this.table.length
    const columns = this.columns.length > 0 ? this.columns[0].length : 0

    return range(0, rows).map(rowIndex =>
      range(0, columns).map(columnIndex =>
        this.table
          .getColumnAt(columnIndex)
          ?.get(rowIndex)
          ?.toString()
      )
    )
  }

  get tableId(): string | undefined {
    return this.styler?.uuid && `T_${this.styler.uuid}`
  }

  get tableStyles(): string | undefined {
    return this.styler?.styles
  }

  get caption(): string | undefined {
    return this.styler?.caption
  }

  get dimensions(): TableDimensions {
    const [headerColumns, dataRowsCheck] = this.index.length
      ? [this.index.length, this.index[0].length]
      : [0, 0]

    const [headerRows, dataColumnsCheck] = this.columns.length
      ? [this.columns.length, this.columns[0].length]
      : [0, 0]

    const [dataRows, dataColumns] = this.data.length
      ? [this.data.length, this.data[0].length]
      : // If there is no data, default to the number of header columns.
        [0, dataColumnsCheck]

    if (
      (dataRows !== 0 && dataRows !== dataRowsCheck) ||
      (dataColumns !== 0 && dataColumns !== dataColumnsCheck)
    ) {
      throw new Error(
        "Table dimensions don't align: " +
          `rows(${dataRows} != ${dataRowsCheck}) OR ` +
          `cols(${dataColumns} != ${dataColumnsCheck})`
      )
    }

    const rows = headerRows + dataRows
    const columns = headerColumns + dataColumns

    return {
      headerRows,
      headerColumns,
      dataRows,
      dataColumns,
      rows,
      columns,
    }
  }

  public getCell(rowIndex: number, columnIndex: number): TableCell {
    const { headerRows, headerColumns, rows, columns } = this.dimensions

    if (rowIndex < 0 || rowIndex >= rows) {
      throw new Error("Row index is out of range.")
    }
    if (columnIndex < 0 || columnIndex >= columns) {
      throw new Error("Column index is out of range.")
    }

    const isBlankCell = rowIndex < headerRows && columnIndex < headerColumns
    const isIndexCell = rowIndex >= headerRows && columnIndex < headerColumns
    const isColumnsCell = rowIndex < headerRows && columnIndex >= headerColumns

    if (isBlankCell) {
      const classNames = ["blank"]
      if (columnIndex > 0) {
        classNames.push(`level${rowIndex}`)
      }
      return {
        type: "blank",
        classNames: classNames.join(" "),
        content: "",
      }
    }

    if (isIndexCell) {
      const dataRowIndex = rowIndex - headerRows

      const uuid = this.styler?.uuid
      const classNames = [
        `row_heading`,
        `level${columnIndex}`,
        `row${dataRowIndex}`,
      ]

      return {
        type: "index",
        id: uuid
          ? `T_${uuid}level${columnIndex}_row${dataRowIndex}`
          : undefined,
        classNames: classNames.join(" "),
        // Table index is stored as is (in the column format).
        content: this.index[columnIndex][dataRowIndex],
      }
    }

    if (isColumnsCell) {
      const dataColumnIndex = columnIndex - headerColumns

      const classNames = [
        `col_heading`,
        `level${rowIndex}`,
        `col${dataColumnIndex}`,
      ]

      return {
        type: "columns",
        classNames: classNames.join(" "),
        content: this.columns[rowIndex][dataColumnIndex],
      }
    }

    const dataRowIndex = rowIndex - headerRows
    const dataColumnIndex = columnIndex - headerColumns

    const uuid = this.styler?.uuid
    const classNames = ["data", `row${dataRowIndex}`, `col${dataColumnIndex}`]

    const content = this.styler?.displayValues
      ? this.styler.displayValues.getCell(rowIndex, columnIndex).content
      : this.data[dataRowIndex][dataColumnIndex]

    return {
      type: "data",
      id: uuid
        ? `T_${uuid}row${dataRowIndex}_col${dataColumnIndex}`
        : undefined,
      classNames: classNames.join(" "),
      content,
    }
  }
}
