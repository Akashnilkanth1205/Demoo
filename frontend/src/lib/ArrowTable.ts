/**
 * @license
 * Copyright 2018-2019 Streamlit Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Table } from "apache-arrow"
import { Map as ImmutableMap } from "immutable"

type CellType = "blank" | "index" | "columns" | "data"

interface Cell {
  classNames: string
  content: string
  id?: string
  type: CellType
}

export default class ArrowTable {
  private readonly dataTable: Table
  private readonly indexTable: Table
  private readonly columnsTable: Table
  private readonly styler?: ImmutableMap<string, any>

  constructor(
    dataBuffer: Uint8Array,
    indexBuffer: Uint8Array,
    columnsBuffer: Uint8Array,
    styler?: any
  ) {
    this.dataTable = Table.from(dataBuffer)
    this.indexTable = Table.from(indexBuffer)
    this.columnsTable = Table.from(columnsBuffer)
    this.styler = styler
  }

  get rows(): number {
    console.log(
      "rows",
      this.indexTable.length,
      this.columnsTable.numCols,
      this.columnsTable.length,
      this.columnsTable.schema.fields
    )
    return this.indexTable.length + this.columnsTable.numCols
  }

  get columns(): number {
    return this.indexTable.numCols + this.columnsTable.length
  }

  get headerRows(): number {
    return this.rows - this.dataRows
  }

  get headerColumns(): number {
    return this.columns - this.dataColumns
  }

  get dataRows(): number {
    return this.dataTable.length
  }

  get dataColumns(): number {
    return this.dataTable.numCols
  }

  get uuid(): string | null {
    return this.styler?.get("uuid", null)
  }

  get caption(): string | null {
    return this.styler?.get("caption", null)
  }

  get styles(): string | null {
    return this.styler?.get("styles", null)
  }

  public getCell = (rowIndex: number, columnIndex: number): Cell => {
    const isBlankCell =
      rowIndex < this.headerRows && columnIndex < this.headerColumns
    const isIndexCell =
      rowIndex >= this.headerRows && columnIndex < this.headerColumns
    const isColumnsCell =
      rowIndex < this.headerRows && columnIndex >= this.headerColumns

    if (isBlankCell) {
      const classNames = ["blank"]
      if (columnIndex > 0) {
        classNames.push("level" + rowIndex)
      }

      return {
        type: "blank",
        classNames: classNames.join(" "),
        content: "",
      }
    } else if (isColumnsCell) {
      const dataColumnIndex = columnIndex - this.headerColumns
      const classNames = [
        "col_heading",
        "level" + rowIndex,
        "col" + dataColumnIndex,
      ]

      return {
        type: "columns",
        classNames: classNames.join(" "),
        content: this.getContent(this.columnsTable, dataColumnIndex, rowIndex),
      }
    } else if (isIndexCell) {
      const dataRowIndex = rowIndex - this.headerRows
      const classNames = [
        "row_heading",
        "level" + columnIndex,
        "row" + dataRowIndex,
      ]

      return {
        type: "index",
        id: `T_${this.uuid}level${columnIndex}_row${dataRowIndex}`,
        classNames: classNames.join(" "),
        content: this.getContent(this.indexTable, dataRowIndex, columnIndex),
      }
    } else {
      const dataRowIndex = rowIndex - this.headerRows
      const dataColumnIndex = columnIndex - this.headerColumns
      const classNames = ["data", "row" + dataRowIndex, "col" + dataColumnIndex]

      return {
        type: "data",
        id: `T_${this.uuid}row${dataRowIndex}_col${dataColumnIndex}`,
        classNames: classNames.join(" "),
        content: this.getContent(this.dataTable, dataRowIndex, dataColumnIndex),
      }
    }
  }

  private getContent = (
    table: Table,
    rowIndex: number,
    columnIndex: number
  ): string => {
    const column = table.getColumnAt(columnIndex)
    if (column === null) {
      return ""
    }
    return column.get(rowIndex)
  }
}
