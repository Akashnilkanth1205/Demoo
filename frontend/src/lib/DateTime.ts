/**
 * @license
 * Copyright 2018-2020 Streamlit Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import moment from "moment-timezone"
import { Datetime } from "autogen/proto"

class DateTimeHandler {
  static moment(date: any, timezone?: string): moment.Moment {
    const nanos = date.datetime || (date.get && date.get("datetime"))
    let datetime = moment(nanos ? this.nanosToMilli(nanos) : date)
    const datetimeTimezone =
      timezone || date.timezone || (date.get && date.get("timezone"))
    return datetimeTimezone
      ? datetime.tz(datetimeTimezone)
      : datetime.tz("UTC").tz(moment.tz.guess(), true)
  }

  static nanosToMilli(nanos: number): number {
    return nanos / 1e6
  }

  static nanosToDuration(nanos: number): Duration {
    return new Duration(nanos / 1e6)
  }

  static dateToString(date: any): string {
    date = this.moment(date)
    let format = "lll z"
    if (date.hour() === 0 && date.minute() === 0 && date.second() === 0) {
      format = "ll z"
    }

    return date.format(format)
  }
}

class Duration {
  private readonly millis: number

  constructor(millis: number) {
    this.millis = millis
  }

  getTime(): number {
    return this.millis
  }
}

export { DateTimeHandler, Duration }
