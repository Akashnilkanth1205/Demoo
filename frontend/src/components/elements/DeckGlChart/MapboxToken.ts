/**
 * @license
 * Copyright 2018-2020 Streamlit Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { SessionInfo } from "lib/SessionInfo"

export const TOKENS_URL = "https://streamlit.io/tokens.json"

/**
 * Exposes a singleton MapboxToken. If the user has specified
 * their own
 */
export class MapboxToken {
  private static token?: string

  public static async get(): Promise<string> {
    if (MapboxToken.token == null) {
      if (SessionInfo.current.userMapboxToken !== "") {
        MapboxToken.token = SessionInfo.current.userMapboxToken
      } else {
        MapboxToken.token = await this.fetchToken(TOKENS_URL, "mapbox")
      }
    }

    return MapboxToken.token
  }

  private static async fetchToken(
    url: string,
    tokenName: string
  ): Promise<string> {
    const rsp = await fetch(url)
    if (!rsp.ok) {
      throw new Error(`${url}: Bad status ${rsp.status}`)
    }

    const json = await rsp.json()
    const token = json[tokenName]
    if (token == null || token === "") {
      throw new Error(`${url}: Missing token "${tokenName}"`)
    }

    return token
  }
}
