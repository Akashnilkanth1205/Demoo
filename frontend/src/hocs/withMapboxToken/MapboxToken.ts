/**
 * @license
 * Copyright 2018-2020 Streamlit Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import axios from "axios"
import { SessionInfo } from "lib/SessionInfo"

/**
 * A remote file that stores user-visible tokens.
 */
export const TOKENS_URL = "https://data.streamlit.io/tokens.json"

export class MapboxToken {
  private static token?: string

  /**
   * Expose a singleton MapboxToken:
   * - If the user specified a token in their streamlit config, return it.
   * - Else, fetch the remote "tokens.json" and return the "mapbox" entry.
   *
   * (The returned value is cached in memory, so the remote resource will
   * only be fetched once per session.)
   */
  public static async get(): Promise<string> {
    if (
      !this.isItRunningLocal() &&
      SessionInfo.current.userMapboxToken === ""
    ) {
      throw new Error(
        "To use this you'll need a Mapbox access token. Please add it to your config."
      )
    }

    if (!MapboxToken.token) {
      if (SessionInfo.current.userMapboxToken !== "") {
        MapboxToken.token = SessionInfo.current.userMapboxToken
      } else {
        MapboxToken.token = await this.fetchToken(TOKENS_URL, "mapbox")
      }
    }

    return MapboxToken.token
  }

  private static isItRunningLocal = (): boolean => {
    const { hostname } = window.location

    return hostname === "localhost" || hostname === "127.0.0.1"
  }

  private static async fetchToken(
    url: string,
    tokenName: string
  ): Promise<string> {
    try {
      const response = await axios.get(url)
      const { mapbox: token } = response.data

      if (token == null || token === "") {
        throw new Error(`Missing token "${tokenName}"`)
      }

      return token
    } catch (e) {
      throw new Error(`${e.message} (${url})`)
    }
  }
}
