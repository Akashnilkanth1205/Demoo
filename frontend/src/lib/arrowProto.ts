/**
 * @license
 * Copyright 2018-2020 Streamlit Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Column, Table } from "apache-arrow"
import { range, unzip } from "lodash"
import { Data, Index, IndexType, Quiver } from "src/lib/Quiver"

export interface DataFrame {
  index: Index
  columns: string[][]
  data: Data
  styler?: Styler
}

interface Schema {
  index_columns: (string | RangeIndex)[]
  columns: SchemaColumn[]
  column_indexes: any[]
}

interface IndexData {
  data: string[]
  type: IndexType
}

interface RangeIndex {
  kind: "range"
  name: string | null
  start: number
  step: number
  stop: number
}

interface SchemaColumn {
  field_name: string
  metadata: Record<string, any> | null
  name: string | null
  numpy_type: string
  pandas_type: string
}

export enum IndexTypes {
  UnicodeIndex = "unicode",
  RangeIndex = "range",
  CategoricalIndex = "categorical",
  MultiIndex = "multiIndex",
  IntervalIndex = "interval[int64]",
  DatetimeIndex = "datetime",
  TimedeltaIndex = "time",
  PeriodIndex = "period[Q-DEC]",
  Int64Index = "int64",
  UInt64Index = "uint64",
  Float64Index = "float64",
}

export interface Styler {
  uuid?: string
  caption?: string
  styles?: string
  displayValues?: any
}

export function parseTable(table: Table): DataFrame {
  const schema = getSchema(table)
  const index = {
    data: getIndex(table, schema),
    type: getIndexType(schema),
  }
  const columns = getColumns(schema)
  const data = {
    data: getData(table, columns),
    type: getColumnType(schema),
  }
  return {
    index,
    columns,
    data,
  }
}

function getSchema(table: Table): Schema {
  const schema = table.schema.metadata.get("pandas")
  if (schema == null) {
    throw new Error("Table schema is missing.")
  }
  return JSON.parse(schema)
}

function getIndex(table: Table, schema: Schema): string[][] {
  return schema.index_columns.map(field => {
    const isRangeIndex = typeof field === "object" && field.kind === "range"
    if (isRangeIndex) {
      const { start, stop, step } = field as RangeIndex
      return range(start, stop, step).map(String)
    }
    const column = table.getColumn(field as string)
    return getColumnData(column).map(String)
  })
}

function getIndexType(schema: Schema): IndexType[] {
  return schema.index_columns.map(indexName => {
    if (isRangeIndex(indexName)) {
      return {
        name: IndexTypes.RangeIndex,
        meta: indexName,
      }
    }

    // Get index column from columns schema.
    const indexColumn = schema.columns.find(
      column => column.field_name === indexName
    )

    // PeriodIndex and IntervalIndex values are kept in `numpy_type` property,
    // the rest in `pandas_type`.
    return {
      name:
        indexColumn?.pandas_type === "object"
          ? indexColumn?.numpy_type
          : indexColumn?.pandas_type,
      meta: indexColumn?.metadata,
    }
  })
}

function getColumns(schema: Schema): string[][] {
  const isMultiIndex = schema.column_indexes.length > 1
  return unzip(
    schema.columns
      .map(column => column.field_name)
      .filter(fieldName => !schema.index_columns.includes(fieldName))
      .map(fieldName =>
        isMultiIndex
          ? JSON.parse(
              fieldName
                .replace(/\(/g, "[")
                .replace(/\)/g, "]")
                .replace(/'/g, '"')
            )
          : [fieldName]
      )
  )
}

function getColumnType(schema: Schema): any {
  return schema.columns
    .filter(column => !schema.index_columns.includes(column.field_name))
    .map(column => column.pandas_type)
}

function getData(table: Table, columns: string[][]): string[][] {
  const rows = table.length
  const cols = columns.length > 0 ? columns[0].length : 0

  return range(0, rows).map(rowIndex =>
    range(0, cols).map(columnIndex =>
      table.getColumnAt(columnIndex)?.get(rowIndex)
    )
  )
}

function getColumnData(column: Column): string[] {
  return range(0, column.length).map(rowIndex => column.get(rowIndex))
}

function isRangeIndex(field: string | RangeIndex): boolean {
  return typeof field === "object" && field.kind === "range"
}

export function concatTables(q1: Quiver, q2: Quiver): DataFrame {
  // Special case if df1 is empty.
  if (q1.data.data.length === 0) {
    return {
      index: q2.index,
      data: q2.data,
      columns: q2.columns,
    }
  }

  // Always return df1 columns
  const { columns } = q1
  const index = concatIndices(q1.index, q2.index)
  const data = concatData(q1.data, q2.data)
  return { index, data, columns }
}

function concatIndices(i1: Index, i2: Index): Index {
  // Special case if i1 is empty.
  if (i1.data.length === 0) {
    return i2
  }

  // Otherwise, make sure the types match.
  if (!sameIndexTypes(i1.type, i2.type)) {
    throw new Error(
      `Cannot concatenate ${JSON.stringify(i1.type)} with ${JSON.stringify(
        i2.type
      )}.`
    )
  }

  return i1.data.reduce(
    (newIndex: Index, indexData: string[], index: number) => {
      const concatenatedIndex = concatIndex(
        indexData,
        i2.data[index],
        i1.type[index]
      )
      newIndex.data.push(concatenatedIndex.data)
      newIndex.type.push(concatenatedIndex.type)
      return newIndex
    },
    { data: [], type: [] }
  )
}

function concatIndex(i1: string[], i2: string[], type: IndexType): IndexData {
  // Special case for RangeIndex.
  if (type.name === IndexTypes.RangeIndex) {
    return concatRangeIndex(i1, i2, type)
  }
  return concatAnyIndex(i1, i2, type)
}

function concatRangeIndex(
  i1: string[],
  i2: string[],
  type: IndexType
): IndexData {
  let newStop = type.meta.stop
  return i2.reduce(
    (newIndex: IndexData) => {
      newIndex.data.push(newStop)
      newStop += type.meta.step
      newIndex.type.meta.stop = newStop
      return newIndex
    },
    { data: i1, type }
  )
}

function concatAnyIndex(
  i1: string[],
  i2: string[],
  type: IndexType
): IndexData {
  const concatenatedIndex = i1.concat(i2)

  // Special case for CategoricalIndex:
  // `num_categories` must be increased.
  if (type.name === IndexTypes.CategoricalIndex) {
    type.meta.num_categories += i2.length
  }

  return { data: concatenatedIndex, type }
}

function concatData(d1: Data, d2: Data): Data {
  // Make sure data types match.
  if (!sameColumnTypes(d1.type, d2.type)) {
    throw new Error(
      `Cannot concatenate ${JSON.stringify(d1.type)} with ${JSON.stringify(
        d2.type
      )}.`
    )
  }

  const numberOfColumns = d1.type.length
  return {
    data: d1.data.concat(
      d2.data.map((data: string[]) => data.slice(0, numberOfColumns))
    ),
    type: d1.type,
  }
}

function sameIndexTypes(t1: IndexType[], t2: IndexType[]): boolean {
  // Make sure both indices have same dimensions.
  if (t1.length !== t2.length) {
    return false
  }

  return t1.every(
    (firstTypeValue: IndexType, index: number) =>
      firstTypeValue.name === t2[index].name
  )
}

function sameColumnTypes(firstType: string[], secondType: string[]): boolean {
  return firstType.every(
    (firstTypeValue: string, index: number) =>
      firstTypeValue === secondType[index]
  )
}
