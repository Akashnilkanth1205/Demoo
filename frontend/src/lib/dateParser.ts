import parser, { Parser, Eat, Tokenizer } from "remark-parse"
import { DateTimeHandler } from "lib/DateTime"
import { Datetime } from "autogen/proto"
function dates() {
  //@ts-ignore
  var Parser = this.Parser
  var tokenizers = Parser.prototype.inlineTokenizers
  var methods = Parser.prototype.inlineMethods

  tokenizers.datetime = tokenizeDatetimes as Tokenizer
  // Run it just before `text`.
  methods.splice(methods.indexOf("text"), 0, "datetime")
}

tokenizeDatetimes.notInLink = true
tokenizeDatetimes.locator = locateDate

function locateDate(value: string, fromIndex: number) {
  return value.indexOf("{@", fromIndex)
}

function tokenizeDatetimes(eat: Eat, value: string, silent: boolean) {
  var match = /{@datetime: (.*)}(.*$)/.exec(value)
  if (match) {
    if (silent) {
      return true
    }

    const date = Datetime.fromObject(JSON.parse(match[1]))
    console.log(JSON.parse(match[1]), date)
    // Reconstruct markdown
    const markdownParts = [
      { type: "inlineCode", value: DateTimeHandler.dateToString(date) },
    ]
    if (match.index) {
      markdownParts.unshift({
        type: "text",
        value: value.substring(0, match.index),
      })
    }
    if (match[2]) {
      markdownParts.push({ type: "text", value: match[2] })
    }

    return eat(value)({
      type: "root",
      children: markdownParts,
    })
  }
}

export default dates
