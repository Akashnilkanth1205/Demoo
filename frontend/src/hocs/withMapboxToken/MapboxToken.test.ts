/**
 * @license
 * Copyright 2018-2020 Streamlit Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import axios from "axios"
import { SessionInfo } from "lib/SessionInfo"
import AxiosMockAdapter from "axios-mock-adapter"
import { MapboxToken, TOKENS_URL } from "hocs/withMapboxToken/MapboxToken"

function setSessionInfoWithMapboxToken(userMapboxToken: string): void {
  SessionInfo.current = new SessionInfo({
    sessionId: "mockSessionId",
    streamlitVersion: "sv",
    pythonVersion: "pv",
    installationId: "iid",
    authorEmail: "ae",
    maxCachedMessageAge: 2,
    commandLine: "cl",
    userMapboxToken: userMapboxToken,
  })
}

describe("MapboxToken", () => {
  let axiosMock: AxiosMockAdapter

  beforeEach(() => {
    window.location.hostname = "localhost"
    axiosMock = new AxiosMockAdapter(axios)
    setSessionInfoWithMapboxToken("")
  })

  afterEach(() => {
    axiosMock.restore()
    MapboxToken["token"] = undefined
    SessionInfo["singleton"] = undefined
  })

  test("Returns cached token if defined", async () => {
    MapboxToken["token"] = "cached"

    await expect(MapboxToken.get()).resolves.toEqual("cached")
  })

  test("Returns userMapboxToken if non-empty", async () => {
    const userToken = "nonEmptyToken"

    setSessionInfoWithMapboxToken(userToken)
    await expect(MapboxToken.get()).resolves.toEqual(userToken)

    // The token should also be cached.
    expect(MapboxToken["token"]).toEqual(userToken)
  })

  test("Fetches remote token if userMapboxToken is empty", async () => {
    const remoteToken = "remoteMapboxToken"

    axiosMock.onGet(TOKENS_URL).reply(200, { mapbox: remoteToken })

    await expect(MapboxToken.get()).resolves.toEqual(remoteToken)

    // The token should also be cached.
    expect(MapboxToken["token"]).toEqual(remoteToken)
  })

  test("Errors if remote token is missing", async () => {
    axiosMock.onGet(TOKENS_URL).replyOnce(200, { ohNo: "noTokenHere" })

    await expect(MapboxToken.get()).rejects.toEqual(
      new Error(`Missing token "mapbox" (${TOKENS_URL})`)
    )

    // No cached token after failure.
    expect(MapboxToken["token"]).toBeUndefined()

    axiosMock.onGet(TOKENS_URL).replyOnce(404, {})
    await expect(MapboxToken.get()).rejects.toEqual(
      new Error(`Request failed with status code 404 (${TOKENS_URL})`)
    )

    // No cached token after failure.
    expect(MapboxToken["token"]).toBeUndefined()
  })

  it("Errors if localhost and missing token", async () => {
    delete window.location
    window.location = { hostname: "http://streamlit.io" } as Location
    setSessionInfoWithMapboxToken("")

    await expect(MapboxToken.get()).rejects.toEqual(
      new Error(
        "To use this you'll need a Mapbox access token. Please add it to your config."
      )
    )
  })
})
