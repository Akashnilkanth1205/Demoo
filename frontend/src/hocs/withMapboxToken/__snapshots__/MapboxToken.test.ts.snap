// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MapboxToken Errors if not hello.py and missing token 1`] = `
[Error: 
            To use this you'll need a Mapbox access token. Please add it to your config.
            
            To get a token for yourself, create an account at
            <a href="https://mapbox.com">https://mapbox.com</a>. It's free! (for moderate usage levels) See
            <a href="https://docs.streamlit.io/cli.html#view-all-config-options">our documentation</a> for more
            info on how to set config options.
            ]
`;

exports[`MapboxToken Errors if not localhost and missing token 1`] = `
[Error: 
            To use this you'll need a Mapbox access token. Please add it to your config.
            
            To get a token for yourself, create an account at
            <a href="https://mapbox.com">https://mapbox.com</a>. It's free! (for moderate usage levels) See
            <a href="https://docs.streamlit.io/cli.html#view-all-config-options">our documentation</a> for more
            info on how to set config options.
            ]
`;
